### ISSUE TITLE
Slow android Dispatchers.Main init

### ISSUE THREAD

**Issue by Tolriq** [Kotlin/kotlinx.coroutines#878]
It seems Android Main dispatcher initialization / discovery can be quite slow currently.

On a Nokia 8 device my application startup time went from 250ms to 500ms (on release mode). I know that 250ms does not sound much but it's pretty much 2 times slower startup time and it does impact the users as they often want the fastest possible startup time for this kind of application. And I still have to see on older devices if the impact is 250ms or linear with the device power.

See attached flamechart (It's from the debug app with full trace profiling so the times are insanely slow but was necessary to find the issue).

![coroutine flame chart](https://user-images.githubusercontent.com/932609/49688233-2e059880-fb0f-11e8-853d-af07f914b5d0.png)

Is there a way to bypass that slow factory discovery and statically create my own Main dispatcher on Android to avoid that slowdown?

**Comment by JakeWharton** [comment]
ServiceLoader should definitely be avoided on Android. I filed https://issuetracker.google.com/issues/120436373 recently to rewrite it in release builds.

**Comment by elizarov** [comment]
This is really strange. I see lot of time spent on reading manifest, but we do not use manifest to initialize `Dispatchers.Main`. 

Is there any chance you can compare loading an app that uses `Dispatchers.Main` (dynamically loaded) vs static reference to `Handler(Looper.getMainLooper()).asCoroutineDispatcher()` to see where exactly this additional time comes from?

**Comment by Tolriq** [comment]
Does ```Handler(Looper.getMainLooper()).asCoroutineDispatcher()``` provide the exact same thing? 

https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt seems to do a lot more.

Looking for temporary workaround as I can't really rollback now :)

**Comment by qwwdfsad** [comment]
>What would be the options that we have to avoid ServiceLoader?

`Class.forName` :sigh:

**Comment by qwwdfsad** [comment]
>Does Handler(Looper.getMainLooper()).asCoroutineDispatcher() provide the exact same thing?

More or less, modulo [`createAsync` ](https://developer.android.com/reference/kotlin/android/os/Handler#createasync_13)

**Comment by Tolriq** [comment]
Did a quick test by duplicating the original code and making the Main public.
(I don't know why but I was not able to use the usual package trick to access still some Kotlin mysteries)

The gain is more in the 100/125ms seems I have something else to find, Maybe 100ms is less important and can wait for the R8 solution, providing a simple public getter for the rare people that will notice and care about it in the meantime could be enough?

**Comment by elizarov** [comment]
What and why is exactly slow in `ServiceLoader` on Android? 

What if we write our own (compatible) implementation that uses `ClassLoader.getResources` to get the corresponding `META-INF/service/xxx` files and parses them. Would it be fast or just as slow?

**Comment by qwwdfsad** [comment]
In the attached flamegraph most of the time is spent in signed JAR verification.
Though I am not sure the same issue arises in release-mode dex: there is no separate JAR and it can be verified only once on startup, not during service loading.

**Comment by qwwdfsad** [comment]
It appears that JAR is verified all the time even in dex.
We can write our own ServiceLoader and avoid verification (it is a matter of one `JarFile` constructor parameter) if it gives a real performance boost.

But I am not sure if we **should**: it will add maintenance burden, it may not work on some devices (especially on new androidx where even default SL breaks in some versions).


Example of how to read service name without verification on Android:
```
val service = "META-INF/services/kotlinx.coroutines.internal.MainDispatcherFactory"
val url = this.javaClass.classLoader.getResources(service).toList()[0]
val string = url.toString()
val separatorIndex = string.indexOf('!')
val pathToJar = string.substring(0, separatorIndex)
val entry = string.substring(separatorIndex + 2, string.length)
val file = JarFile(pathToJar.substring("jar:file:/".length), false)
BufferedReader(InputStreamReader(file.getInputStream(ZipEntry(entry)), "UTF-8")).useLines {
    it.forEach { println(it) } // prints AndroidDispatcherFactory
}
```

**Comment by Tolriq** [comment]
Maybe I was not clear but the given values where from production app in release mode and heavily optimized ;)

If you do nothing I would highly appreciate to have a public getter in https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt for those who care about the delay in application init.

Currently it seems Kotlin prevent using a package trick to access the internal Main, leading to necessary code duplication just to pass the var public and avoid the issue.

Maybe add another public alias with a clear name / Kdoc about it's usage not recommended?

**Comment by Tolriq** [comment]
One more detail, I got some ANR in prod due to that too.

Probably due to accessing file from main thread on a busy file system.

```
"main" prio=5 tid=1 Runnable
  | group="main" sCount=0 dsCount=0 flags=0 obj=0x71e769a8 self=0xb303a000
  | sysTid=8148 nice=0 cgrp=default sched=0/0 handle=0xb6e794a4
  | state=R schedstat=( 688284631 7931945867 493 ) utm=46 stm=22 core=0 HZ=100
  | stack=0xbe505000-0xbe507000 stackSize=8MB
  | held mutexes= "mutator lock"(shared held)
  at java.util.HashMap.put (HashMap.java:611)
  at sun.security.util.ManifestDigester.<init> (ManifestDigester.java:170)
  at java.util.jar.JarVerifier.processEntry (JarVerifier.java:290)
- locked <0x0c34e60a> (a byte[])
  at java.util.jar.JarVerifier.update (JarVerifier.java:229)
  at java.util.jar.JarFile.initializeVerifier (JarFile.java:374)
  at java.util.jar.JarFile.getInputStream (JarFile.java:441)
- locked <0x06592e7b> (a java.util.jar.JarFile)
  at libcore.io.ClassPathURLStreamHandler$ClassPathURLConnection.getInputStream (ClassPathURLStreamHandler.java:177)
  at java.net.URL.openStream (URL.java:1058)
  at java.util.ServiceLoader.parse (ServiceLoader.java:305)
  at java.util.ServiceLoader.-wrap0 (ServiceLoader.java)
  at java.util.ServiceLoader$LazyIterator.hasNextService (ServiceLoader.java:358)
  at java.util.ServiceLoader$LazyIterator.hasNext (ServiceLoader.java:402)
  at java.util.ServiceLoader$1.hasNext (ServiceLoader.java:488)
  at kotlin.a.u.a (_Collections.kt:1132)
  at kotlin.a.u.f (_Collections.kt:1165)
  at kotlin.a.u.e (_Collections.kt:1156)
  at kotlinx.coroutines.bx.<clinit> (Dispatchers.kt:96)
  at kotlinx.coroutines.ax.b (Dispatchers.kt:53)
  at com.genimee.android.yatse.b.a.b.b.<init> (MainThreadBus.kt:14)
  at com.genimee.android.yatse.b.b.a.<clinit> (ServiceLocator.kt:24)
  at org.leetzone.android.yatsewidget.YatseApplication.<clinit> (YatseApplication.kt:390)
  at java.lang.Class.newInstance (Native method)
  at android.app.Instrumentation.newApplication (Instrumentation.java:1102)
  at android.app.Instrumentation.newApplication (Instrumentation.java:1087)
  at android.app.LoadedApk.makeApplication (LoadedApk.java:983)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5720)
  at android.app.ActivityThread.-wrap1 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1657)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6499)
  at java.lang.reflect.Method.invoke (Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:442)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)
```

**Comment by LouisCAD** [comment]
There's a solution to have the Main dispatcher being automatically injected before `Application.onCreate()` is called. It is by using a `ContentProvider` (which are eagerly initialized) and taking advantage of the default manifest placeholder for applicationId so there's no clash between apps.
This needs to have the android artifact be published as an aar though. This could also be done with a separate artifact, provided there's a public API to make it possible.

You can see some example on something I done for this: https://github.com/LouisCAD/Splitties/blob/e77c909585f1b6d457af0fe18655e4794434ce50/initprovider/README.md

**Comment by JakeWharton** [comment]
But... why? That mechanism provides no value here except breaking
multi-process use.

**Comment by LouisCAD** [comment]
@JakeWharton Because it requires no additional code in consumer apps. in other words, it can fix the issue for the default process in a convenient way (having nothing to do apart from updating).

The current slow path can still be used for multiprocess use (so it doesn't break anything). Such processes are usually not the processes where the main UI is run anyway, so cold start time doesn't matter as much.

**Comment by JakeWharton** [comment]
That still doesn't answer why...

You know what else doesn't require additional code but is lazy and works in
all processes? Regular class loading.

The content provider hack is only for getting access to a Context. Regular
static initialization is superior in every way when that isn't needed.

**Comment by LouisCAD** [comment]
The current design doesn't allow static initialization because `Dispatchers.Main` is shared with non android kotlinx.coroutines, so that's why there was a ServiceLoader in the first place, and that's why I proposing an alternative to initialize it. I think I now answered why, do you agree?

About the ContentProvider hack, it is not just for getting access to a Context but equally to run before `Application.onCreate` is called, hence why it can fit this use case.

**Comment by Tolriq** [comment]
Before talking about how, let's talk about if it should be done as @qwwdfsad seems to think it's not :(

When talking about how will come the do 'we do file access on main thread' (as possible solution proposed by @qwwdfsad do too), and does the solution needs to be global or optional? (Meaning should the faster path replace the Dispatchers.Main or be something else that won't interfere with other implementations and open a lot more solutions).

**Comment by qwwdfsad** [comment]
I see multiple solutions here and I'm especially interested in @JakeWharton opinion on that.

1) Do nothing, wait until R8 starts desugaring service loaders into static access. Very tempting for `kotlinx.coroutines`, resolves the issue in the long term but does not help users on older toolchains.

2) Use `Class.forName(AndroidDispatcherFactory fqn)`. It is the fastest solution, but it can have a weird interaction with e.g. TestDispatcher (#749) and adds a lot of maintenance burden. We should keep in mind such behaviour when implementing new features, it is testable only when `kotlinx-coroutines-android` artifact is present in the classpath, it does not resolve the same issue with `CoroutineExceptionHandler`. I generally don't like this solution.

3)  Write our own `ServiceLoader`. According to my experiments, it speeds up `load` by an order of magnitude. Yes, it is still a file read from the main thread, but this effect is negligible, and it cuts down 90% of startup footprint. I don't see any problems with omitted JAR verification because we can statically reject any services that are not part of `kotlinx.coroutines` and/or we can make verification an opt-in mechanism for environments where JARs from untrusted sources is the real issue (and AFAIK Android is not one of these environments).
Though I am not sure whether custom `ServiceLoader` will always work on every Android runtime, we can always fallback to the default `ServiceLoader`. This solution looks the most beneficial for me and we probably can even implement it in `1.1.0` release.

**Comment by Tolriq** [comment]
> Yes, it is still a file read from the main thread, but this effect is negligible

Let me disagree on that, was hard to reach top 3% of Play Store ANR (And can't go further due to Android 8+ bug) and any file access on main thread as small as it can be, can have impact on low end phones and block for many seconds before getting access due to some low level locks.
So yes 3) cuts a lot of the issue in nominal cases, but does not solve all cases, so as you said does it worth the maintenance cost if it's not 100% reliable and full solution?

**Comment by LouisCAD** [comment]
@qwwdfsad Is there a reason you don't talk about adding a public (possibly experimental) API to inject an instance of the main dispatcher?
This would allow us to instantiate it ourselves, statically, from host application code, bypassing the usual lookup that unfortunately does I/O on main thread. I think it would be a great solution until R8 "desugars" ServiceLoaders (which could be distant in time, and ANRs…).

**Comment by qwwdfsad** [comment]
Yes, I've explained them in https://github.com/Kotlin/kotlinx.coroutines/issues/810#issuecomment-439927255

**Comment by LouisCAD** [comment]
@qwwdfsad To me, the solution I mentioned, using the `ContentProvider` hack, is the only solution that can be implemented as a workaround without drawbacks or public API changes while waiting for R8 desugar support. It falls back to serviceloader on non default process, but most apps are single process as I said previously, and it's not worse than what we currently have anyway.

I can try to make a PR for this.

**Comment by Tolriq** [comment]
Got confirmation that 1) won't be planned before quite some time as priority is R8 stability before those things and handling the jar change directly at OS is out of possibility minSdk 29 would be 10 years away :)

For my needs, I'll go to prod with duplication of the handler and not using the Dispatchers.Main.

In what cases / at what point is the CoroutineExceptionHandler serviceLoader started?

**Comment by sellmair** [comment]
> Yes, I've explained them in [#810 (comment)](https://github.com/Kotlin/kotlinx.coroutines/issues/810#issuecomment-439927255)

So the argument is that allowing to change the default Dispatchers publicly and globally is bad for any core library like coroutines because this could lead to third-party libraries conflicts, right?

Would this argument also hold, if one would just allow global configuration once and further changes are disallowed by the framework (by throwing a Runtime Exception that explains that re-assignment of the main dispatcher is not possible)

Isn't this conceptionally the same as defining the service class in META-INF/services/...?: 
Anyone could define his own service there. 

The tradeoff would be compile-time safety vs performance here. I personally could live with the Runtime Exception when loading two 'UI coroutine libraries'.

**Comment by LouisCAD** [comment]
Good point you made Sebastian, allowing once is a good option, and since
we're in Kotlin 1.3, this API could be marked as exeprimental with error
level, preventing unaware usages, and likely limiting it to just
kotlinx.coroutines android artifact (if using the ContentProvider hack).
However, instead of throwing, a Boolean could be returned, so only first
set returns true, and subsequent, if any, wouldn't crash the app by default
but would just be ignored.

**Comment by qwwdfsad** [comment]
>Isn't this conceptionally the same as defining the service class in META-INF/services/..., in regards of conflicts?

It is not. Service loader (and actual service interface) mechanism is purely internal. We do not allow third-parties to implement custom main dispatchers, we do not provide any backward compatibility guarantees and there are chances we even forbid implementations not known to our service loader.

Set-once is still a source of confusion, non-determinism and (as any other global mutable state provided by a library or a runtime) eventually will end up with reordering your project dependencies along with reflection hacks until you get the desired behaviour. Yes, it is better than "set anything anytime", but it still has the same flaws. 

>The tradeoff would be compile-time safety vs performance here. 

There is no tradeoffs in your solution. Tradeoff between safety and performance is to start using `Class.forName` or rolling our own ServiceLoader, not opening dispatcher for global mutation.

> I personally could live with the Runtime Exception when loading two 'UI coroutine libraries'.

Yes, everyone could. Until they have tens of dependencies that have 7 versions of 3 conflicting logging frameworks*


What you (and others) are proposing is a good solution for a local project or for a library which is used only within a single company, especially when you've already seen similar pattern in `RxJavaPlugins`. But it is not good enough for a core library and eventually will harm the ecosystem.
Moreover, global settable dispatcher is not really helpful: for tests there is #749, if you really need settable dispatches -- use dependeny injection (you don't set global statics in order to mock a dependency, do you?).


\* No jokes: https://github.com/eBay/Spark/blob/master/core/src/main/scala/org/apache/spark/Logging.scala#L162
Another good example is implicit shutdown hook dependencies.

**Comment by elizarov** [comment]
@Tolriq There is a public non-experimental API for you: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-android/kotlinx.coroutines.android/android.os.-handler/as-coroutine-dispatcher.html

You can statically create a coroutine dispatcher from your handler and then inject it statically throughout your code. As long as you do not explicitly use `Dispatchers.Main` there is no dynamic loading (it is loaded only on the actual access to this property).

**Comment by Tolriq** [comment]
@elizarov yes but as confirmed by @qwwdfsad it does not do exactly the same, I currently have copied the https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt made Main public and inject that and it works nicely, just sad to have to copy the file to have the var public. (Usually I would use the package trick to access it, but it seems for file level variable this does not work)

For the moment I can use that workaround, but it means I may need to update my code copy when I update coroutines as result might be unpredictable if I forget.

Do you have the answer to the question about CoroutineExceptionHandler  serviceloader call? Trying to find why Crashlytics now impact my startup time even if initialized on background thread and want to know if it can be linked or not.

**Comment by elizarov** [comment]
@Tolriq 

> In what cases / at what point is the CoroutineExceptionHandler serviceLoader started?

It happens the first time this `handleCoroutineExceptionImpl` function is invoked:
https://github.com/Kotlin/kotlinx.coroutines/blob/d6a5a399d1724ff56bbb285b25df071dbc98b715/core/kotlinx-coroutines-core/src/CoroutineExceptionHandlerImpl.kt#L21
It happens if and only if there is uncaught exception and it was not handled by your code. For Android it means that your app is about to crash.

>  I currently have copied the https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt

The only thing you _might_ need to copy from there is this `Looper.asHandler` function:
https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt#L61
However, depending on your API level it could be much simpler implemented.  You don't need the rest of that code, because `Handler.asCoroutineDispatcher` is public, or I don't understand your use-case.

**Comment by Tolriq** [comment]
Thanks for the details about CoroutineExceptionHandler I'll search more what can be the cause :( I fear it's due to all the threads Crashlytics start at the same time as coroutine init + https://github.com/Kotlin/kotlinx.coroutines/issues/840 :(

About the code duplication I admit that I stopped looking too deep after seeing 2 internals and failed with the trick and directly used the hammer. Your right I just need to duplicate the asHandler and as I'm api 16 -> 28 I can simplify that part too.
Wanted a quick workaround in prod as I discovered the delay a little too late. Will do that better for next release.

**Comment by diegomontoya** [comment]
Is there a bug in this line of asHandler?
`
    if (Build.VERSION.SDK_INT >= 28) {
        // TODO compile against API 28 so this can be invoked without reflection.
        val factoryMethod = Handler::class.java.getDeclaredMethod("createAsync", Looper::class.java)
        return factoryMethod.invoke(null, this) as Handler
    }
`
For api >= 28, async boolean is ignore since createAsync is always called.

**Comment by Ornolfr** [comment]
Is this going to be fixed some time in the future? It's causing massive lags:

```
D/StrictMode: StrictMode policy violation; ~duration=640 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:241)
        at java.io.File.isDirectory(File.java:845)
        at java.io.File.toURI(File.java:728)
        at libcore.io.ClassPathURLStreamHandler.<init>(ClassPathURLStreamHandler.java:52)
        at dalvik.system.DexPathList$Element.maybeInit(DexPathList.java:702)
        at dalvik.system.DexPathList$Element.findResource(DexPathList.java:729)
        at dalvik.system.DexPathList.findResources(DexPathList.java:526)
        at dalvik.system.BaseDexClassLoader.findResources(BaseDexClassLoader.java:174)
        at java.lang.ClassLoader.getResources(ClassLoader.java:839)
        at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:349)
        at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:402)
        at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:488)
        at kotlin.collections.CollectionsKt___CollectionsKt.toCollection(_Collections.kt:1145)
        at kotlin.collections.CollectionsKt___CollectionsKt.toMutableList(_Collections.kt:1178)
        at kotlin.collections.CollectionsKt___CollectionsKt.toList(_Collections.kt:1169)
        at kotlinx.coroutines.internal.MainDispatcherLoader.loadMainDispatcher(MainDispatchers.kt:15)
        at kotlinx.coroutines.internal.MainDispatcherLoader.<clinit>(MainDispatchers.kt:10)
        at kotlinx.coroutines.Dispatchers.getMain(Dispatchers.kt:55)
```

**Comment by Tolriq** [comment]
Fix that is currently in the work will mitigate that but won't fix.

Still disk IO on main thread.

**Comment by Ornolfr** [comment]
@Tolriq thank you for your quick response. I used this third party library and it seems to have fixed the problem for me:
https://github.com/LouisCAD/Splitties/tree/master/lifecycle-coroutines
Here's the commit with their fix:
https://github.com/LouisCAD/Splitties/commit/816c4a6fcfcc4e20ac4bef839dad13b88a72c705

Is it also just mitigating and not fixing the actual issue?

**Comment by Tolriq** [comment]
@LouisCAD version does not touches IO on main thread so if better for Android apps.

But it does not use asynchronous handler. I have not perf tested the difference but I suppose that it's better with the proper handler. And I prefer to be more close to original implementation in all cases.

For my usage I have copied private asHandler and use it (This version require compileSDK 28.

```
fun Looper.asHandler(async: Boolean): Handler {
    if (!async) {
        return Handler(this)
    }

    if (Build.VERSION.SDK_INT >= 28) {
        return Handler.createAsync(this)
    }

    val constructor: Constructor<Handler>
    try {
        constructor = Handler::class.java.getDeclaredConstructor(Looper::class.java, Handler.Callback::class.java, Boolean::class.javaPrimitiveType)
    } catch (ignored: NoSuchMethodException) {
        return Handler(this)
    }
    return constructor.newInstance(this, null, true)
}
```

Then ```Looper.getMainLooper().asHandler(async = true).asCoroutineDispatcher("Main")``` to instanciate it.

(I actually have duplicated the whole file and not use the android library to be sure to crash and never call the wrong one) But it works the same.

Would be nice that asHandler was public but it seems it's not currently an option.

**Comment by LouisCAD** [comment]
You can PR Splitties to use async Handler by branching from the develop
branch. If it's of sufficient quality, I'll merge it and then release it on
jcenter next weekend.

**Comment by LouisCAD** [comment]
@Tolriq I just released version 3.0.0-alpha03 of Splitties, where `Dispatchers.MainAndroid` is now backed by an async `Handler`.

**Comment by Tolriq** [comment]
Thanks but I tend to not use too much external libraries and have implemented my solution since mid December :) This is more for @Ornolfr 

Let's hope an official no disk on, main thread solution is made one day.

**Comment by Tolriq** [comment]
For those who will find this, the "fix" mitigate the issue, but this is still disk IO on main thread.

On Android, this is still an issue and can still cause slowdown / ANR on busy devices. (Obviously no more than without that patch, but still an issue for those who seek 100% ANR free sessions)

**Comment by qwwdfsad** [comment]
It is a very isolated IO for a single file that should be in a disk cache anyway.

Feel free to open a new issue if you **still** experience ANR after `1.2.0-alpha-2`; 
There are ways to speed up that loading even more, but only if there exists real evidence that the current solution is not enough

**Comment by Tolriq** [comment]
I do Android apps since 8 years every single disk access on main thread can and will trigger an ANR or slow down at some point, flash is slow, 50$ phones have even more bad quality flash.

I no more use the official dispatcher but my own to address that, so won't be able to give "evidence" about something well known.

I had to provide Google with a patch for their Billing library because on recent updates Google Play did a simple file access on service connection, leading to tons of ANRs.

Most users do not check their ANRs stats or vitals and it's really hard to spot on tests without forcing the device to have a busy IO queue, so you probably won't have other users reporting the issue for a while, but it will still be present.

I totally understand that Android is not the only / main target and that 0,x% of slow down and 0,0X% of ANR is not a priority. But it's important that for those who matters, they know there's still a disk IO and the cause.

**Comment by LouisCAD** [comment]
I just submitted a feature request for the Android Emulator to allow simulating these slow I/O performances while the storage is already busy and risks triggering an ANR because a file has been naively read from the main thread: https://issuetracker.google.com/issues/129900564

Hopefully, this will at some point allow us to reproduce these situations more easily and consistently.

**Comment by elizarov** [comment]
What we really need in a R8 feature that would desugar `ServiceLoader` usage into static class initialization based on the `META-INF/services` files that are found when building project.

**Comment by JakeWharton** [comment]
Preliminary support landed last week. Haven't tested it yet, but I plan to
this week.

**Comment by elizarov** [comment]
@JakeWharton We also need to figure out how we shall write the code so that:

```
if (newR8isAvailable) useRegualarServiceLoader() else useOurOwnWorkAround()
```

**Comment by JakeWharton** [comment]
Detecting you're being processed by R8 is pretty easy. Detecting what
version of R8... that is less easy.

**Comment by nak411** [comment]
> It is a very isolated IO for a single file that should be in a disk cache anyway.

Why was this issue closed?  @qwwdfsad Are there any plans to address this so developers can develop with strict mode on?
I recently discovered this and can use some of the workarounds mentioned above to get around it but doing IO on the main thread is never appropriate for android. https://stackoverflow.com/questions/55584194/android-kotlin-coroutine-crashing-on-strict-mode

**Comment by elizarov** [comment]
@nak411 Can you, please, create a separate issue with a reproducer code from that SO question, so that we have an appropriate place to discuss and track it.

**Comment by qwwdfsad** [comment]
No I/O on the Main thread since 1.3.3, we gave up on using `ServiceLoader`
